File: backend/app/schemas/reporting.py
"""Pydantic schemas for the Reporting and Analytics API."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Per-institution stats
# ---------------------------------------------------------------------------


class InstitutionStats(BaseModel):
    institution_id: uuid.UUID
    institution_code: str
    institution_name: str
    institution_type: str
    faculty_count: int
    department_count: int
    programme_count: int
    module_count: int
    audit_run_count: int
    evidence_file_count: int
    knowledge_indexed: bool
    qdrant_collection: str | None


# ---------------------------------------------------------------------------
# Knowledge index status per collection
# ---------------------------------------------------------------------------


class KnowledgeIndexEntry(BaseModel):
    institution_code: str
    academic_year: str
    ikp_version: str
    collection: str
    indexed: bool
    chunk_count: int | None


# ---------------------------------------------------------------------------
# Dashboard
# ---------------------------------------------------------------------------


class DashboardResponse(BaseModel):
    institution_count: int
    faculty_count: int
    department_count: int
    programme_count: int
    module_count: int
    audit_run_count: int
    completed_audit_count: int
    failed_audit_count: int
    evidence_file_count: int
    knowledge_index_status: list[KnowledgeIndexEntry]
    by_institution: list[InstitutionStats]
    generated_at: datetime
    is_admin_view: bool


# ---------------------------------------------------------------------------
# Faculty / Programme / Module summaries
# ---------------------------------------------------------------------------


class FacultySummaryResponse(BaseModel):
    faculty_id: uuid.UUID
    faculty_name: str
    institution_code: str
    department_count: int
    programme_count: int
    module_count: int


class ProgrammeSummaryResponse(BaseModel):
    programme_id: uuid.UUID
    programme_name: str
    programme_code: str
    nqf_level: int | None
    faculty_name: str
    institution_code: str
    module_count: int
    audit_run_count: int


class ModuleSummaryResponse(BaseModel):
    module_id: uuid.UUID
    module_name: str
    module_code: str
    academic_year: str
    programme_name: str
    institution_code: str
    audit_run_count: int
    evidence_file_count: int
    latest_audit_status: str | None


# ---------------------------------------------------------------------------
# Compliance summary
# ---------------------------------------------------------------------------


class ComplianceSummaryResponse(BaseModel):
    institution_code: str
    total_modules: int
    audited_modules: int
    compliant_count: int
    at_risk_count: int
    non_compliant_count: int
    unaudited_count: int
    compliance_rate_pct: float

File: frontend/src/types/reporting.ts
export interface KnowledgeIndexEntry {
  institution_code: string;
  academic_year: string;
  ikp_version: string;
  collection: string;
  indexed: boolean;
  chunk_count: number | null;
}

export interface InstitutionStats {
  institution_id: string;
  institution_code: string;
  institution_name: string;
  institution_type: string;
  faculty_count: number;
  department_count: number;
  programme_count: number;
  module_count: number;
  audit_run_count: number;
  evidence_file_count: number;
  knowledge_indexed: boolean;
  qdrant_collection: string | null;
}

export interface DashboardResponse {
  institution_count: number;
  faculty_count: number;
  department_count: number;
  programme_count: number;
  module_count: number;
  audit_run_count: number;
  completed_audit_count: number;
  failed_audit_count: number;
  evidence_file_count: number;
  knowledge_index_status: KnowledgeIndexEntry[];
  by_institution: InstitutionStats[];
  generated_at: string;
  is_admin_view: boolean;
}

export interface FacultySummaryResponse {
  faculty_id: string;
  faculty_name: string;
  institution_code: string;
  department_count: number;
  programme_count: number;
  module_count: number;
}

export interface ProgrammeSummaryResponse {
  programme_id: string;
  programme_name: string;
  programme_code: string;
  nqf_level: number | null;
  faculty_name: string;
  institution_code: string;
  module_count: number;
  audit_run_count: number;
}

export interface ModuleSummaryResponse {
  module_id: string;
  module_name: string;
  module_code: string;
  academic_year: string;
  programme_name: string;
  institution_code: string;
  audit_run_count: number;
  evidence_file_count: number;
  latest_audit_status: string | null;
}

export interface ComplianceSummaryResponse {
  institution_code: string;
  total_modules: number;
  audited_modules: number;
  compliant_count: number;
  at_risk_count: number;
  non_compliant_count: number;
  unaudited_count: number;
  compliance_rate_pct: number;
}
